feat(template): Add `--notes` flag

Bhargav-InfraCloud opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

Added a new flag `--notes` to the `helm template` command, with the default set to `false`.  
If enabled, it will print the rendered `NOTES.txt` at the end of the manifests.  
If `--render-subchart-notes` (this flag already exists on the `template` command, as it is inherited from the `install` command) is also enabled, the output will include the subcharts’ notes as well.

Updated the `install` command’s `--render-subchart-notes` flag description to clarify that, in the case of the `template` command, it is only valid when the `--notes` flag is enabled.  
Note: the `template` command uses most of the `install` command’s flags, including `--render-subchart-notes`.

Updated `action.renderResources()` to format the root chart’s notes, and the subcharts’ notes if the `--render-subchart-notes` flag is enabled, and return them as a string. The output depends on whether it is:

- **Dry run**  
  - Notes are prefixed with the source filename and separated with `---`.  
  - For `helm install` / `helm upgrade`, the `--dry-run` flag enables this behavior.  
  - The `helm template` command is considered a dry run by default.  
  - Example:
    ```
    foo NOTES HERE

    bar NOTES HERE

    baz NOTES HERE

    qux NOTES HERE
    ```

- **Not a dry run**  
  - Only the root chart’s notes are rendered by default.  
  - If the `--render-subchart-notes` flag is enabled, the subcharts’ notes are also included.  
  - Each note is prefixed with the source filename and separated with `---`.  
  - Example:
    ```
    ---
    # Source: foo/templates/NOTES.txt
    foo NOTES HERE
    ---
    # Source: foo/charts/bar/templates/NOTES.txt
    bar NOTES HERE
    ---
    # Source: foo/charts/baz/templates/NOTES.txt
    baz NOTES HERE
    ---
    # Source: foo/charts/bar/charts/qux/templates/NOTES.txt
    qux NOTES HERE
    ```

In both cases, the notes are sorted based on the following conditions:  
1. Depth  
2. Alphabetical order (since the notes file name is always the same, the sorting applies to the chart/subchart name).

Added tests for the following scenarios:
 - helm template:
   1. On the chart without notes or subcharts
   2. `--notes` on chart without notes or subcharts
   3. `--render-subchart-notes` on chart without notes or subcharts
   4. `--notes --render-subchart-notes` on chart without notes or subcharts
   5. On the chart with notes without subcharts
   6. `--notes` on chart with notes without subcharts
   7. `--render-subchart-notes` on chart with notes without subcharts
   8. `--notes --render-subchart-notes` on chart with notes without subcharts
   9. On the chart with notes and subchart
   10. `--notes` on chart with notes and subchart
   11. `--render-subchart-notes` on the chart with  notes and subchart
   12. `--notes --render-subchart-notes` on chart  with notes and subchart
   13. `--help`
 - helm install:
   1. On the chart without notes or subcharts
   2. `--dry-run` on chart without notes or subcharts
   3. `--render-subchart-notes` on chart without notes or subcharts
   4. `--dry-run --render-subchart-notes` on chart without notes or subcharts
   5. On the chart with notes without subcharts
   6. `--dry-run` on chart with notes without subcharts
   7. `--render-subchart-notes` on chart with notes without subcharts
   8. `--dry-run --render-subchart-notes` on chart with notes without subcharts
   9. On a chart with notes and subcharts
   10. `--dry-run` on chart with notes and  subcharts
   11. `--render-subchart-notes` on chart with  notes and subcharts
   12. `--dry-run --render-subchart-notes` on chart  with notes and subcharts
   13. `--help`
 - helm upgrade:
   1. On the chart without notes or subcharts
   2. `--dry-run` on chart without notes or subcharts
   3. `--render-subchart-notes` on chart without notes or subcharts
   4. `--dry-run --render-subchart-notes` on chart without notes or subcharts
   5. On the chart with notes without subcharts
   6. `--dry-run` on chart with notes without subcharts
   7. `--render-subchart-notes` on chart with notes without subcharts
   8. `--dry-run --render-subchart-notes` on chart with notes without subcharts
   9. On the chart with notes and subcharts
   10. `--dry-run` on chart with notes and subcharts
   11. `--render-subchart-notes` on chart with  notes and subcharts
   12. `--dry-run --render-subchart-notes` on  chart with notes and subcharts
   13. `--help`

Fixes #6901

**Special notes for your reviewer**:

- The method `(*Configuration).renderResources()` from `action` package was returning the notes string even when there is an error. This doesn't make any sense as upper layers though assigning it to `rel.Info.Notes`, but never use it as the run itself exits with an error. So, I replaced it with empty strings. No tests are failing. My guess is some old and stale code. But if you think it is needed, let me know.

**If applicable**:
- [x] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility


---

## Testing

Follow these steps to test the new and existing notes flags:

### 1. Clone and Build

- Clone the forked Helm repository:
    ```
    git clone https://github.com/Bhargav-InfraCloud/helm
    ```
- Switch to the feature branch:
    ```
    git checkout bhargav-ravuri/6901/add_notes_flag_to_helm_template_command
    ```
- Build the Helm binary:
    ```
    make build
    ```
- Copy the built binary to a directory in your `$PATH`:
    ```
    cp ./bin/helm /usr/local/bin/
    ```

### 2. Create Nested Helm Charts

- Create a temporary directory and nested charts with custom `NOTES.txt` files:
    ```sh
    cd $(mktemp -d)
    helm create foo && echo "foo TEXT HERE" > foo/templates/NOTES.txt
    cd foo/charts
    helm create bar && echo "bar TEXT HERE" > bar/templates/NOTES.txt
    helm create baz && echo "baz TEXT HERE" > baz/templates/NOTES.txt
    cd bar/charts
    helm create qux && echo "qux TEXT HERE" > qux/templates/NOTES.txt
    cd ../../../..
    ```

### 3. Test Commands

Run the following commands on the nested Helm charts you created in the previous step to verify the behavior of the `--notes` and `--render-subchart-notes` flags.

#### helm template

1. `helm template luffy foo`
    - Expected: Helm template output without any notes (from the current chart or subcharts).
    - Output: [1.1.helm-template-luffy-foo.txt](https://github.com/user-attachments/files/22661545/1.1.helm-template-luffy-foo.txt)

2. `helm template luffy foo --notes`
    - Expected: Helm template output with only the current chart's notes.
    - Output: [1.2.helm-template-luffy-foo-with-flag-notes.txt](https://github.com/user-attachments/files/22661547/1.2.helm-template-luffy-foo-with-flag-notes.txt)

3. `helm template luffy foo --render-subchart-notes`
    - Expected: Helm template output without any notes (from the current chart or subcharts), since subchart notes are only rendered when `--notes` is enabled.
    - Output: [1.3.helm-template-luffy-foo-with-flag-render-subchart-notes.txt](https://github.com/user-attachments/files/22661548/1.3.helm-template-luffy-foo-with-flag-render-subchart-notes.txt)

4. `helm template luffy foo --notes --render-subchart-notes`
    - Expected: Helm template output with both the current chart's and subcharts' notes, in order (first by depth, then alphabetically).
    - Output: [1.4.helm-template-luffy-foo-with-flag-notes-and-render-subchart-notes.txt](https://github.com/user-attachments/files/22661550/1.4.helm-template-luffy-foo-with-flag-notes-and-render-subchart-notes.txt)

#### helm install

1. `helm install luffy foo`
    - Expected: Helm install output with the current chart's notes, as `--notes` is enabled by default in `helm install`.
    - Output: [2.1.helm-install-luffy-foo.txt](https://github.com/user-attachments/files/22661551/2.1.helm-install-luffy-foo.txt)

2. `helm install luffy foo --dry-run`
    - Expected: Helm install dry-run output with the current chart's notes, as `--notes` is enabled by default in `helm install`.
    - Output: [2.2.helm-install-luffy-foo-with-flag-dry-run.txt](https://github.com/user-attachments/files/22661552/2.2.helm-install-luffy-foo-with-flag-dry-run.txt)

3. `helm install luffy foo --render-subchart-notes`
    - Expected: Helm install output with the current chart's and subcharts' notes, as `--notes` is enabled by default in `helm install`, in order (first by depth, then alphabetically).
    - Output: [2.3.helm-install-luffy-foo-with-flag-render-subchart-notes.txt](https://github.com/user-attachments/files/22661553/2.3.helm-install-luffy-foo-with-flag-render-subchart-notes.txt)

4. `helm install luffy foo --dry-run --render-subchart-notes`
    - Expected: Helm install dry-run output with the current chart's and subcharts' notes, as `--notes` is enabled by default in `helm install`, in order (first by depth, then alphabetically).
    - Output: [2.4.helm-install-luffy-foo-with-flag-dry-run-and-render-subchart-notes.txt](https://github.com/user-attachments/files/22661555/2.4.helm-install-luffy-foo-with-flag-dry-run-and-render-subchart-notes.txt)

#### helm upgrade

1. `helm upgrade luffy foo`
    - Expected: Helm upgrade output with the current chart's notes, as `--notes` is enabled by default in `helm upgrade`.
    - Output: [3.1.helm-upgrade-luffy-foo.txt](https://github.com/user-attachments/files/22661559/3.1.helm-upgrade-luffy-foo.txt)

2. `helm upgrade luffy foo --dry-run`
    - Expected: Helm upgrade dry-run output with the current chart's notes, as `--notes` is enabled by default in `helm upgrade`.
    - Output: [3.2.helm-upgrade-luffy-foo-with-flag-dry-run.txt](https://github.com/user-attachments/files/22661562/3.2.helm-upgrade-luffy-foo-with-flag-dry-run.txt)

3. `helm upgrade luffy foo --render-subchart-notes`
    - Expected: Helm upgrade output with the current chart's and subcharts' notes, as `--notes` is enabled by default in `helm upgrade`, in order (first by depth, then alphabetically).
    - Output: [3.3.helm-upgrade-luffy-foo-with-flag-render-subchart-notes.txt](https://github.com/user-attachments/files/22661567/3.3.helm-upgrade-luffy-foo-with-flag-render-subchart-notes.txt)

4. `helm upgrade luffy foo --dry-run --render-subchart-notes`
    - Expected: Helm upgrade dry-run output with the current chart's and subcharts' notes, as `--notes` is enabled by default in `helm upgrade`, in order (first by depth, then alphabetically).
    - Output: [3.4.helm-upgrade-luffy-foo-with-flag-dry-run-and-render-subchart-notes.txt](https://github.com/user-attachments/files/22661570/3.4.helm-upgrade-luffy-foo-with-flag-dry-run-and-render-subchart-notes.txt)

Bhargav-InfraCloud committed:
feat(template): Add --notes flag

Added a new flag `--notes` to the `helm template`
command, with the default set to `false`.
If enabled, it will print the rendered `NOTES.txt`
at the end of the manifests.
If `--render-subchart-notes` (this flag already
exists on the `template` command, as it is
inherited from the `install` command) is also
enabled, the output will include the subcharts'
notes as well.

Updated the `install` command's
`--render-subchart-notes` flag description to
clarify that, in the case of the `template`
command, it is only valid when the `--notes` flag
is enabled.
Note: the `template` command uses most of the
`install` command's flags, including
`--render-subchart-notes`.

Updated `action.renderResources()` to format the
root chart's notes, and the subcharts' notes if
the `--render-subchart-notes` flag is enabled, and
return them as a string. The output depends on
whether it is:

- **Dry run**
  - Notes are prefixed with the source filename
    and separated with `---`.
  - For `helm install` / `helm upgrade`, the
    `--dry-run` flag enables this behavior.
  - The `helm template` command is considered a
    dry run by default.
  - Example:
    ```
    foo NOTES HERE

    bar NOTES HERE

    baz NOTES HERE

    qux NOTES HERE
    ```

- **Not a dry run**
  - Only the root chart's notes are rendered by
    default.
  - If the `--render-subchart-notes` flag is
    enabled, the subcharts' notes are also
    included.
  - Each note is prefixed with the source filename
    and separated with `---`.
  - Example:
    ```
    ---
    # Source: foo/templates/NOTES.txt
    foo NOTES HERE
    ---
    # Source: foo/charts/bar/templates/NOTES.txt
    bar NOTES HERE
    ---
    # Source: foo/charts/baz/templates/NOTES.txt
    baz NOTES HERE
    ---
    # Source: foo/charts/bar/charts/qux/templates/NOTES.txt
    qux NOTES HERE
    ```

In both cases, the notes are sorted based on the
following conditions:
1. Depth
2. Alphabetical order (since the notes file name
   is always the same, the sorting applies to the
   chart/subchart name).

Added tests for the following scenarios:
 - helm template:
   1. On the chart without notes or subcharts
   2. `--notes` on chart without notes or
      subcharts
   3. `--render-subchart-notes` on chart without
      notes or subcharts
   4. `--notes --render-subchart-notes` on chart
      without notes or subcharts
   5. On the chart with notes without subcharts
   6. `--notes` on chart with notes without
      subcharts
   7. `--render-subchart-notes` on chart with
      notes without subcharts
   8. `--notes --render-subchart-notes` on chart
      with notes without subcharts
   9. On the chart with notes and subchart
   10. `--notes` on chart with notes and subchart
   11. `--render-subchart-notes` on the chart with
       notes and subchart
   12. `--notes --render-subchart-notes` on chart
       with notes and subchart
   13. `--help`
 - helm install:
   1. On the chart without notes or subcharts
   2. `--dry-run` on chart without notes or
      subcharts
   3. `--render-subchart-notes` on chart without
      notes or subcharts
   4. `--dry-run --render-subchart-notes` on chart
      without notes or subcharts
   5. On the chart with notes without subcharts
   6. `--dry-run` on chart with notes without
      subcharts
   7. `--render-subchart-notes` on chart with
      notes without subcharts
   8. `--dry-run --render-subchart-notes` on chart
      with notes without subcharts
   9. On a chart with notes and subcharts
   10. `--dry-run` on chart with notes and
       subcharts
   11. `--render-subchart-notes` on chart with
       notes and subcharts
   12. `--dry-run --render-subchart-notes` on
       chart with notes and subcharts
   13. `--help`
 - helm upgrade:
   1. On the chart without notes or subcharts
   2. `--dry-run` on chart without notes or
      subcharts
   3. `--render-subchart-notes` on chart without
      notes or subcharts
   4. `--dry-run --render-subchart-notes` on chart
      without notes or subcharts
   5. On the chart with notes without subcharts
   6. `--dry-run` on chart with notes without
      subcharts
   7. `--render-subchart-notes` on chart with
      notes without subcharts
   8. `--dry-run --render-subchart-notes` on chart
      with notes without subcharts
   9. On the chart with notes and subcharts
   10. `--dry-run` on chart with notes and
       subcharts
   11. `--render-subchart-notes` on chart with
       notes and subcharts
   12. `--dry-run --render-subchart-notes` on
       chart with notes and subcharts
   13. `--help`

Closes #6901

Signed-off-by: Bhargav Ravuri <bhargav.ravuri@infracloud.io>

benoittgt commented:
Hello @Bhargav-InfraCloud,  

Thanks for the PR! I tested it locally and it works as expected.  
Here are the commands I used for testing:  

```sh
./bin/helm template myrelease test-chart
./bin/helm template myrelease test-chart --notes
./bin/helm template myrelease test-chart --notes --render-subchart-notes
./bin/helm template myrelease test-chart --render-subchart-notes
./bin/helm install myrelease test-chart --dry-run
./bin/helm install myrelease test-chart --dry-run --render-subchart-notes
```

I'm just not sure this behavior https://github.com/helm/helm/issues/6901#issuecomment-2652091864 will work as _notes_  are not valid yaml. Piping the output directly into `kubectl apply` can cause issues.  

Example:  

```sh
./bin/helm template myrelease test-chart --notes > manifests.yaml
kubectl apply --dry-run=client -f manifests.yaml
```

Results:  

```text
serviceaccount/myrelease-subchart created (dry run)
serviceaccount/myrelease-test-chart created (dry run)
service/myrelease-subchart created (dry run)
service/myrelease-test-chart created (dry run)
deployment.apps/myrelease-subchart created (dry run)
deployment.apps/myrelease-test-chart created (dry run)
pod/myrelease-subchart-test-connection created (dry run)
pod/myrelease-test-chart-test-connection created (dry run)
error: error validating "manifests.yaml": error validating data: invalid object to validate; if you choose to ignore these errors, turn validation off with --validate=false
```

So to me, this is more of a developer responsibility :) :  
- either make their notes compatible,  
- or exclude them when generating manifests for `kubectl apply`, no `--*notes`  

Overall, the patch looks solid and thanks for the tests. Could you please fix the conflicts?

Bhargav-InfraCloud commented:
@benoittgt Great! Thanks for reviewing it! I've fixed the conflicts.

> I'm just not sure this behavior https://github.com/helm/helm/issues/6901#issuecomment-2652091864 will work as notes are not valid yaml. Piping the output directly into kubectl apply can cause issues.
> So to me, this is more of a developer responsibility :) :
> 
> - either make their notes compatible,
> - or exclude them when generating manifests for kubectl apply, no --*notes

Agreed. Without the `--notes` flag, the `helm template` output should be the same as before. Users can use it for piping output to `kubectl`.

Bhargav-InfraCloud commented:
> Off the wall idea, but maybe a way to include notes.txt safely so downstream things don't get tripped up?

Since the new behavior is gated behind a new flag `--notes`, it won’t affect existing users.

> `--safe-notes` : prepend `#` before lines of rendered notes.txt?

That's one option, and it would be a small change. But IMO, adding another layer isn't great for UX. Instead, I think users should follow this approach:

```bash
# Render all templates (including NOTES). Output is plain text, not YAML-parseable.
helm template luffy foo --notes
helm template luffy foo --notes --render-subchart-notes

# If YAML parsing is needed:
helm template luffy foo | yq

# To view only the NOTES output:
helm template luffy foo --notes --show-only templates/NOTES.txt
helm template luffy foo --notes --render-subchart-notes --show-only charts/bar/templates/NOTES.txt
```

WDYT?

Bhargav-InfraCloud commented:
@ all, FYI.
- Added testing steps and outputs for common use cases in the description above (unit tests already cover these).
- Rebased on the latest main.

outofrange commented:
Thanks for working on this - I'm mostly using Helm through kustomize or ArgoCD, so I often forget that NOTES.txt are actually a thing :)

I want to suggest writing notes to _stderr_ instead of _stdout_, or making it configurable;
I'd even argue that it's what users could expect as a default, since notes can be seen as _diagnostic output_ for which [POSIX suggests using _stderr_](https://pubs.opengroup.org/onlinepubs/009695399/).

Benefits, at least from my point of view:
- can still be piped to other tools expecting YAML
- `helm template ... --notes | kubectl apply -f -` would even display the notes in the terminal like `helm install` would (although I'm not sure why anyone would directly pipe to apply instead of doing `helm upgrade --install`)
- tools like kustomize could add support for setting `--notes`, also surfacing notes when doing something like `kustomize build --enable-helm . | kubectl apply -f -` (which is quite common, in contrast to directly applying helm templated manifests)
- tools like ArgoCD could simply capture output from _stderr_ for notes, instead of splitting _stdout_, or processing templates twice (which would introduce additional load during the often already quite expensive reconciliation loop)
- since _stdout_ would remain untouched, `--notes=true` could even be the default for `helm template` - either set as default by Helm itself, or configured by users through aliases, functions, env vars, etc... at least I wouldn't see a reason to **not** enable notes when they don't interfere with my other tooling, but might provide valuable information I might miss otherwise

Not sure how I would treat `helm install`; writing to _stderr_ would be justified in this case as well imho, but probably not worth a breaking change. If the output stream for notes would be configurable, `helm template` and `helm install` could use different defaults.

To summarize, my idea for `helm template` is
- write notes to _stderr_
  - if anyone can think of reasons why users would need to have it on stdout: make the output stream configurable through flag, `~/.config/helm` and/or env var
- make it possible to **always** show notes by configuring it through `~/.config/helm` and/or env var

davinkevin commented:
We would love to see this PR merged as we struggle to correctly test the `NOTES.txt`.

scottrigby commented on pkg/cmd/testdata/output/template-help.txt:
```diff
@@ -0,0 +1,86 @@
+
+Render chart templates locally and display the output.
```
same concern. Remove full help output from testdata

scottrigby commented on pkg/cmd/testdata/testcharts/chart-with-notes/templates/rbac/role.yaml:
```diff
@@ -0,0 +1,8 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
we can simplify all of these new test charts dramatically, to only include what we're actually testing, NOTES.txt in the parent and subcharts. Keep as minimal as possible to test out this specific functionality.
